File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        queue_reverse = deque()
        queue_normal = deque()
        queue_normal.append(root)
        queue_reverse.append(root)
        final = []
        is_reverse = False
        
        
        
        if not root:
            return final
        
        
        
        while(queue_normal):
            size = len(queue_normal)
            temp = []
                        
            for i in range(size):
                cur_reverse = queue_reverse.popleft()
                if cur_reverse.right:
                    queue_reverse.append(cur_reverse.right)
                if cur_reverse.left:    
                    queue_reverse.append(cur_reverse.left)
                    
                cur_normal = queue_normal.popleft()    
                if cur_normal.left:
                    queue_normal.append(cur_normal.left)
                if cur_normal.right:
                    queue_normal.append(cur_normal.right)
                if is_reverse:
                    temp.append(cur_reverse.val)
                else:
                    temp.append(cur_normal.val)
            
            
            final.append(temp)
            is_reverse = (not is_reverse)
        return final
                
                    
        

File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        LinkedList<TreeNode> queue = new LinkedList<>();
        if(root != null){
            queue.add(root);
        }
        List<List<Integer>> res = new LinkedList<>();
        boolean isOrderLeft = true;
        
        while(!queue.isEmpty()){
            int size = queue.size();
            List<Integer> temp = new LinkedList<>();
            for(int i = 0; i < size; i++){
                if(isOrderLeft){
                    TreeNode popped = queue.pollFirst();
                    temp.add(popped.val);
                if(popped.left != null){
                    queue.addLast(popped.left);
                }
                if(popped.right != null){
                    queue.addLast(popped.right);
                }
                                       
                }else{
                    TreeNode popped = queue.pollLast();
                    temp.add(popped.val);
                    if(popped.right != null){
                        queue.addFirst(popped.right);
                    }
                    if(popped.left != null){
                        queue.addFirst(popped.left);
                    }
                }

            }
            res.add(temp);
            isOrderLeft = !isOrderLeft;
        }
        return res;
        
        
        
    }
}